Add tolerance as argument

JoerivanEngelen opened the pull request:
Fixes https://github.com/Deltares/numba_celltree/issues/41

Adds ``tolerance`` as argument, to make tolerance configurable. I ran into problems with the HYDAMO dataset, as projected coords are large, which lead to precision errors. I think making tolerance configurable by the user is a neat feature to have.

Huite commented:
I'm in principle okay with adding a tolerance, but the reason I didn't do this before is because it made my head hurt and I didn't have the time to consider things properly. If I remember correctly, one of the problems currently is that the current `TOLERANCE_ON_EDGE` is compared with a cross product value for efficiency -- but this means it's not a clear tolerance in terms of distance, but in terms of the parallelogram area. I've thought about converting it to a meaningful tolerance length, but this requires computing the perpendicular height of the parallelogram, which is more expensive.

Regardless, there's another thing to consider, which is that if you provide an explicit tolerance, I think you create the expectation that it's used everywhere. So I think you'd have to use it in the tree traversal's bounding box comparison as well.

We could take another look at the VTK implementation to get some inspiration -- but I would've guessed I would have done that before.

Anyway, for your specific problem there might be an easier solution, like quantizing your input coordinates. Can you motivate this PR with an issue with some screenshots and the like?

JoerivanEngelen commented:
Update, after a chat with @Huite, we concluded that tolerance for intersect_edges makes not a lot of sense and has a very limited usecase. I removed the tolerance argument and logic from the intersect_edges algorithms, and removed some edge cases in the test, as their behavior was too unpredictable.

jay-hennen commented:
OK I'll come at this with the naive point of view. Why is it a good idea at all to attempt to compensate for floating point precision errors via a tolerance? It seems like creating situations where computations or comparisons only remain consistent in the presence of this tolerance is only a recipe for confusion and frustration down the line.

Huite commented:
## GEOS

GEOS requires a tolerance all the same in these cases; looking in the GEOS issue quickly gives an example (shown here via shapely):

```python
from shapely.geometry import Point, LineString

line = LineString([(0.0, 2.0), (2.0, 0.0)])
point = Point(0.4, 1.6)
point.intersects(line)  # False
line.distance(point)  # 7.850462293418876e-17
```

See comment:
https://github.com/libgeos/geos/issues/1207#issuecomment-2555717041

> In the future the topological predicates may support a distance tolerance, which would solve this problem as well.

Huite commented:
It seems like VTK makes a somewhat similar distinction as I did (numerical versus actual spatial tolerance); `findCellsAlongLine` takes a tolerance argument, but `findCell` doesn't!
https://vtk.org/doc/nightly/html/classvtkCellTreeLocator.html#abc057bb9ec328fc3e7af934270ccd9cc

Huite commented:
I think we're making good progress.

We should probably make a note somewhere of what these tolerances mean; i.e. they should really only be used for floating point issues, although technically for the edge cell tree, you could purposely use it find a point close-ish to an edge. But you'll get no guarantee that it's actually the closest!  -- it'll be the first that it found with an acceptable tolerance.

JoerivanEngelen reviewed (commented):
Thanks for the review, I've improved based on your comments, also implemented a way to estimate a default tolerance.
